ansible/roles/test/files/ptftests/fast-reboot.py

```diff
@@ -265,23 +460,63 @@ def runTest(self):
 
         self.log("Wait until ASIC stops")
         self.timeout(self.task_timeout, "DUT hasn't stopped in %d seconds" % self.task_timeout)
-        no_routing_start = self.check_stop()
+        no_routing_start, upper_replies = self.check_stop()
         self.cancel_timeout()
 
         self.log("ASIC was stopped, Waiting until it's up. Stop time: %s" % str(no_routing_start))
         self.timeout(self.task_timeout, "DUT hasn't started to work for %d seconds" % self.task_timeout)
-        no_routing_stop = self.check_start()
+        no_routing_stop, _ = self.check_start()
         self.cancel_timeout()
+
+        for thr, q in self.ssh_jobs:
+            q.put('quit')
+            thr.join()
+
+        no_cp_replies = self.extract_no_cpu_replies(upper_replies)
+
         self.log("ASIC works again. Start time: %s" % str(no_routing_stop))
 
         self.log("Downtime was %s" % str(no_routing_stop - no_routing_start))
-
-        self.assertTrue(no_routing_stop - no_routing_start < self.limit, "Downtime must be less then %s seconds" % self.test_params['fast_reboot_limit'])
+        self.log("Reboot time was %s" % str(no_routing_stop - self.reboot_start))
+        self.log("Number replies when control plane was down: %d Expected: %d" % (no_cp_replies, self.nr_vl_pkts))
+
+        self.fails['dut'] = set()
+        if no_routing_stop - no_routing_start > self.limit:
+            self.fails['dut'].add("Downtime must be less then %s seconds" % self.test_params['fast_reboot_limit'])
+        if no_routing_stop - self.reboot_start > datetime.timedelta(seconds=self.test_params['graceful_limit']):
+            self.fails['dut'].add("Fast-reboot cycle must be less then graceful limit %s seconds" % self.test_params['graceful_limit'])
+        if no_cp_replies < 0.95 * self.nr_vl_pkts:
+            self.fails['dut'].add("Dataplane didn't route to all servers, when control-plane was down: %d vs %d" % (no_cp_replies, self.nr_vl_pkts))
+
+        is_good = True
+        errors = ""
+        for name, fails in self.fails.items():
+            if len(fails) > 0:
+                is_good = False
+            for fail in fails:
+                self.log("FAILED:%s:%s" % (name, fail))
+                errors += "FAILED:%s:%s\n" % (name, fail)
+
+        self.assertTrue(is_good, "Something went wrong. Please check output below.\n" + errors)
+
+    def extract_no_cpu_replies(self, arr):
+      """
+      This function tries to extract number of replies from dataplane, when control plane is non working
+      """
+      # remove all tail zero values
+      non_zero = filter(lambda x : x > 0, arr)
+
+      # check that last value is different from previos
+      if len(non_zero) > 1 and non_zero[-1] < non_zero[-2]:
+          return non_zero[-2]
+      else:
+          return non_zero[-1]
```

Comment:
maybe it is more robust just to find out the 90% of the data for non zero. Meaning 90% percentage of time, the you have the reachability to the n host, and the n should be equal to nr_vl_pkts.